File: funnyvote/app/src/test/java/com/heaton/funnyvote/ui/createvote/CreateVoteActivityPresenterTest.java
package com.heaton.funnyvote.ui.createvote;

import com.heaton.funnyvote.data.VoteData.VoteDataRepository;
import com.heaton.funnyvote.data.VoteData.VoteDataSource;
import com.heaton.funnyvote.data.user.UserDataRepository;
import com.heaton.funnyvote.data.user.UserDataSource;
import com.heaton.funnyvote.database.Option;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.database.VoteData;

import junit.framework.Assert;

import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anyList;
import static org.mockito.Matchers.anyMap;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class CreateVoteActivityPresenterTest {
    private static User user;
    private static final String voteCode = "CODE_123";
    private static final VoteData voteData = new VoteData();
    private static List<Option> optionList = new ArrayList<>();

    @Mock
    private UserDataRepository userDataRepository;
    @Mock
    private VoteDataRepository voteDataRepository;
    @Mock
    private CreateVoteContract.ActivityView activityView;
    @Mock
    private CreateVoteContract.SettingFragmentView settingFragmentView;
    @Mock
    private CreateVoteContract.OptionFragmentView optionFragmentView;

    private CreateVoteActivityPresenter presenter;

    @Captor
    private ArgumentCaptor<VoteDataSource.GetVoteDataCallback> getVoteDataCallbackArgumentCaptor;

    @Captor
    private ArgumentCaptor<UserDataSource.GetUserCallback> getUserCallbackArgumentCaptor;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);

        user = mock(User.class);
        when(user.getUserName()).thenReturn("Heaton");
    }

    @Test
    public void createPresenter_setsThePresenterToView() {
        // Get a reference to the class under test
        presenter = new CreateVoteActivityPresenter(voteDataRepository, userDataRepository
                , activityView, optionFragmentView, settingFragmentView);

        // Then the presenter is set to the view
        verify(activityView).setPresenter(presenter);
        presenter.setOptionFragmentView(optionFragmentView);
        presenter.setSettingFragmentView(settingFragmentView);
    }

    @Test
    public void getUserAndInitialDefaultView() {
        presenter = new CreateVoteActivityPresenter(voteDataRepository, userDataRepository
                , activityView, optionFragmentView, settingFragmentView);
        presenter.start();
        verify(userDataRepository).getUser(getUserCallbackArgumentCaptor.capture(), eq(false));
        getUserCallbackArgumentCaptor.getValue().onResponse(user);
    }

    @Test
    public void submitCreateVoteSuccessAndShowVoteDetail() {
        presenter = new CreateVoteActivityPresenter(voteDataRepository, userDataRepository
                , activityView, optionFragmentView, settingFragmentView);
        presenter.setVoteSettings(voteData);

        optionList.clear();
        for (long i = 0; i < 2; i++) {
            Option option = new Option();
            option.setTitle("OPTION_CODE_" + i);
            option.setId(i);
            option.setCount(0);
            optionList.add(option);
        }
        presenter.setOptionList(optionList);
        presenter.getVoteSettings().setTitle("TITLE_1");
        presenter.getVoteSettings().setAuthorCode("AUTHOR_1");
        presenter.getVoteSettings().setMaxOption(1);
        presenter.getVoteSettings().setMinOption(1);
        presenter.getVoteSettings().setIsUserCanAddOption(false);
        presenter.getVoteSettings().setIsCanPreviewResult(false);
        presenter.getVoteSettings().setIsNeedPassword(false);
        presenter.getVoteSettings().setSecurity(VoteData.SECURITY_PUBLIC);
        presenter.getVoteSettings().setEndTime(System.currentTimeMillis() + 3000 * 86400 * 1000);
        when(settingFragmentView.getFinalVoteSettings(voteData)).thenReturn(voteData);

        presenter.submitCreateVote();
        verify(activityView).showLoadingCircle();
        verify(voteDataRepository).createVote(any(VoteData.class), anyList()
                , any(File.class), getVoteDataCallbackArgumentCaptor.capture());
        getVoteDataCallbackArgumentCaptor.getValue().onVoteDataLoaded(voteData);
        verify(activityView).showHintToast(anyInt());
        verify(activityView).hideLoadingCircle();
        verify(activityView).IntentToVoteDetail(voteData);
    }

    @Test
    public void submitCreateVoteRemoteFailureAndShowErrorToast() {
        presenter = new CreateVoteActivityPresenter(voteDataRepository, userDataRepository
                , activityView, optionFragmentView, settingFragmentView);
        presenter.setVoteSettings(voteData);

        optionList.clear();
        for (long i = 0; i < 2; i++) {
            Option option = new Option();
            option.setTitle("OPTION_CODE_" + i);
            option.setId(i);
            option.setCount(0);
            optionList.add(option);
        }
        presenter.setOptionList(optionList);
        presenter.getVoteSettings().setTitle("TITLE_1");
        presenter.getVoteSettings().setAuthorCode("AUTHOR_1");
        presenter.getVoteSettings().setMaxOption(1);
        presenter.getVoteSettings().setMinOption(1);
        presenter.getVoteSettings().setIsUserCanAddOption(false);
        presenter.getVoteSettings().setIsCanPreviewResult(false);
        presenter.getVoteSettings().setIsNeedPassword(false);
        presenter.getVoteSettings().setSecurity(VoteData.SECURITY_PUBLIC);
        presenter.getVoteSettings().setEndTime(System.currentTimeMillis() + 3000 * 86400 * 1000);
        when(settingFragmentView.getFinalVoteSettings(voteData)).thenReturn(voteData);

        presenter.submitCreateVote();
        verify(activityView).showLoadingCircle();
        verify(voteDataRepository).createVote(any(VoteData.class), anyList()
                , any(File.class), getVoteDataCallbackArgumentCaptor.capture());
        getVoteDataCallbackArgumentCaptor.getValue().onVoteDataNotAvailable();
        verify(activityView).showHintToast(anyInt());
        verify(activityView).hideLoadingCircle();
        verify(activityView, never()).IntentToVoteDetail(voteData);
    }

    @Test
    public void submitCreateVoteLocalFailureAndShowErrorDialog() {
        presenter = new CreateVoteActivityPresenter(voteDataRepository, userDataRepository
                , activityView, optionFragmentView, settingFragmentView);
        presenter.setVoteSettings(voteData);

        for (long i = 0; i < 2; i++) {
            Option option = new Option();
            option.setTitle("OPTION_CODE_" + i);
            option.setId(i);
            option.setCount(0);
            optionList.add(option);
        }
        presenter.setOptionList(optionList);
        //Only no title case
        presenter.getVoteSettings().setTitle("");
        presenter.getVoteSettings().setAuthorCode("AUTHOR_1");
        presenter.getVoteSettings().setMaxOption(1);
        presenter.getVoteSettings().setMinOption(1);
        presenter.getVoteSettings().setIsUserCanAddOption(false);
        presenter.getVoteSettings().setIsCanPreviewResult(false);
        presenter.getVoteSettings().setIsNeedPassword(false);
        presenter.getVoteSettings().setSecurity(VoteData.SECURITY_PUBLIC);
        presenter.getVoteSettings().setEndTime(System.currentTimeMillis() + 3000 * 86400 * 1000);
        when(settingFragmentView.getFinalVoteSettings(voteData)).thenReturn(voteData);

        presenter.submitCreateVote();
        verify(activityView).showLoadingCircle();
        verify(voteDataRepository, never()).createVote(any(VoteData.class), anyList()
                , any(File.class), getVoteDataCallbackArgumentCaptor.capture());
        verify(activityView).hideLoadingCircle();
        verify(activityView).showCreateVoteError(anyMap());
    }

    @Test
    public void addAndReviseAndRemoveOptionUpdateToView() {
        presenter = new CreateVoteActivityPresenter(voteDataRepository, userDataRepository
                , activityView, optionFragmentView, settingFragmentView);
        long optionId = presenter.addNewOption();
        presenter.addNewOption();
        presenter.addNewOption();
        presenter.reviseOption(optionId, "newText");
        String equalText = "";
        for (Option option : presenter.getOptionList()) {
            if (option.getId() == optionId) {
                equalText = option.getTitle();
            }
        }
        Assert.assertEquals(equalText, "newText");
        verify(optionFragmentView, times(3)).refreshOptions();
        int optionNumber = presenter.getOptionList().size();
        presenter.removeOption(optionId);
        verify(optionFragmentView, times(4)).refreshOptions();
        Assert.assertEquals(presenter.getOptionList().size(), optionNumber - 1);
    }

    @Test
    public void updateVoteSecurityAndEndTimeAndTitleAndUpdateToView() {
        presenter = new CreateVoteActivityPresenter(voteDataRepository, userDataRepository
                , activityView, optionFragmentView, settingFragmentView);
        presenter.updateVoteSecurity(VoteData.SECURITY_PRIVATE);
        Assert.assertEquals(presenter.getVoteSettings().getSecurity(), VoteData.SECURITY_PRIVATE);
        presenter.updateVoteEndTime(System.currentTimeMillis()
                - CreateVoteActivityPresenter.Companion.getDEFAULT_END_TIME() * 1000 * 86400);
        verify(activityView).showHintToast(anyInt());
        presenter.updateVoteEndTime(System.currentTimeMillis()
                + (CreateVoteActivityPresenter.Companion.getDEFAULT_END_TIME_MAX() + 10) * 1000 * 86400);
        verify(activityView).showHintToast(anyInt());
        presenter.updateVoteEndTime(System.currentTimeMillis()
                + CreateVoteActivityPresenter.Companion.getDEFAULT_END_TIME_MAX() * 1000 * 86400);
        verify(settingFragmentView).setUpVoteSettings(any(VoteData.class));
        presenter.updateVoteImage(new File("test"));
        Assert.assertNotNull(presenter.getVoteSettings().getImageFile());
        presenter.updateVoteTitle("title");
        Assert.assertEquals(presenter.getVoteSettings().getTitle(),"title");
    }
}

File: funnyvote/app/src/test/java/com/heaton/funnyvote/ui/createvote/CreateVoteActivityPresenterTest.kt
package com.heaton.funnyvote.ui.createvote

import com.heaton.funnyvote.any
import com.heaton.funnyvote.capture
import com.heaton.funnyvote.data.VoteData.VoteDataRepository
import com.heaton.funnyvote.data.VoteData.VoteDataSource
import com.heaton.funnyvote.data.user.UserDataRepository
import com.heaton.funnyvote.data.user.UserDataSource
import com.heaton.funnyvote.database.Option
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.database.VoteData
import com.heaton.funnyvote.ui.createvote.CreateVoteContract.*
import junit.framework.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.*
import org.mockito.Mockito.*
import java.io.File
import java.util.*

class CreateVoteActivityPresenterTest {

    @Mock
    private lateinit var userDataRepository: UserDataRepository
    @Mock
    private lateinit var voteDataRepository: VoteDataRepository
    @Mock
    private lateinit var activityView: CreateVoteContract.ActivityView
    @Mock
    private lateinit var settingFragmentView: CreateVoteContract.SettingFragmentView
    @Mock
    private lateinit var optionFragmentView: CreateVoteContract.OptionFragmentView

    private lateinit var presenter: CreateVoteActivityPresenter

    @Captor
    private lateinit var getVoteDataCallbackArgumentCaptor: ArgumentCaptor<VoteDataSource.GetVoteDataCallback>

    @Captor
    private lateinit var getUserCallbackArgumentCaptor: ArgumentCaptor<UserDataSource.GetUserCallback>

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)

        user = mock(User::class.java)
        `when`(user.getUserName()).thenReturn("Heaton")
    }

    @Test
    fun createPresenter_setsThePresenterToView() {
        // Get a reference to the class under test
        presenter = CreateVoteActivityPresenter(voteDataRepository, userDataRepository, activityView, optionFragmentView, settingFragmentView)

        // Then the presenter is set to the view
        verify(activityView).setPresenter(presenter)
    }

    @Test
    fun getUserAndInitialDefaultView() {
        presenter = CreateVoteActivityPresenter(voteDataRepository, userDataRepository, activityView, optionFragmentView, settingFragmentView)
        presenter.start()
        verify(userDataRepository).getUser(capture(getUserCallbackArgumentCaptor), ArgumentMatchers.eq(false))
        getUserCallbackArgumentCaptor.value.onResponse(user)
    }

    @Test
    fun submitCreateVoteSuccessAndShowVoteDetail() {
        presenter = CreateVoteActivityPresenter(voteDataRepository, userDataRepository, activityView, optionFragmentView, settingFragmentView)
        presenter.voteSettings = voteData

        optionList.clear()
        for (i in 0..1) {
            val option = Option()
            option.title = "OPTION_CODE_$i"
            option.id = i.toLong()
            option.count = 0
            optionList.add(option)
        }
        presenter.optionList = optionList
        presenter.voteSettings.title = "TITLE_1"
        presenter.voteSettings.authorCode = "AUTHOR_1"
        presenter.voteSettings.maxOption = 1
        presenter.voteSettings.minOption = 1
        presenter.voteSettings.isUserCanAddOption = false
        presenter.voteSettings.isCanPreviewResult = false
        presenter.voteSettings.isNeedPassword = false
        presenter.voteSettings.security = VoteData.SECURITY_PUBLIC
        presenter.voteSettings.endTime = System.currentTimeMillis() + 3000 * 86400 * 1000
        `when`(settingFragmentView.getFinalVoteSettings(voteData)).thenReturn(voteData)

        presenter.submitCreateVote()
        verify<ActivityView>(activityView).showLoadingCircle()
        verify(voteDataRepository).createVote(any<VoteData>(), ArgumentMatchers.anyList(), any<File>(), capture(getVoteDataCallbackArgumentCaptor))
        getVoteDataCallbackArgumentCaptor.value.onVoteDataLoaded(voteData)
        verify<ActivityView>(activityView).showHintToast(ArgumentMatchers.anyInt())
        verify<ActivityView>(activityView).hideLoadingCircle()
        verify<ActivityView>(activityView).IntentToVoteDetail(voteData)
    }

    @Test
    fun submitCreateVoteRemoteFailureAndShowErrorToast() {
        presenter = CreateVoteActivityPresenter(voteDataRepository, userDataRepository, activityView, optionFragmentView, settingFragmentView)
        presenter.voteSettings = voteData

        optionList.clear()
        for (i in 0..1) {
            val option = Option()
            option.title = "OPTION_CODE_$i"
            option.id = i.toLong()
            option.count = 0
            optionList.add(option)
        }
        presenter.optionList = optionList
        presenter.voteSettings.title = "TITLE_1"
        presenter.voteSettings.authorCode = "AUTHOR_1"
        presenter.voteSettings.maxOption = 1
        presenter.voteSettings.minOption = 1
        presenter.voteSettings.isUserCanAddOption = false
        presenter.voteSettings.isCanPreviewResult = false
        presenter.voteSettings.isNeedPassword = false
        presenter.voteSettings.security = VoteData.SECURITY_PUBLIC
        presenter.voteSettings.endTime = System.currentTimeMillis() + 3000 * 86400 * 1000
        `when`(settingFragmentView.getFinalVoteSettings(voteData)).thenReturn(voteData)

        presenter.submitCreateVote()
        verify<ActivityView>(activityView).showLoadingCircle()
        verify(voteDataRepository).createVote(any<VoteData>(), ArgumentMatchers.anyList()
                , any<File>(), capture(getVoteDataCallbackArgumentCaptor))
        getVoteDataCallbackArgumentCaptor.value.onVoteDataNotAvailable()
        verify<ActivityView>(activityView).showHintToast(ArgumentMatchers.anyInt())
        verify<ActivityView>(activityView).hideLoadingCircle()
        verify<ActivityView>(activityView, never()).IntentToVoteDetail(voteData)
    }

    @Test
    fun submitCreateVoteLocalFailureAndShowErrorDialog() {
        presenter = CreateVoteActivityPresenter(voteDataRepository, userDataRepository, activityView, optionFragmentView, settingFragmentView)
        presenter.voteSettings = voteData

        for (i in 0..1) {
            val option = Option()
            option.title = "OPTION_CODE_$i"
            option.id = i.toLong()
            option.count = 0
            optionList.add(option)
        }
        presenter.optionList = optionList
        //Only no title case
        presenter.voteSettings.title = ""
        presenter.voteSettings.authorCode = "AUTHOR_1"
        presenter.voteSettings.maxOption = 1
        presenter.voteSettings.minOption = 1
        presenter.voteSettings.isUserCanAddOption = false
        presenter.voteSettings.isCanPreviewResult = false
        presenter.voteSettings.isNeedPassword = false
        presenter.voteSettings.security = VoteData.SECURITY_PUBLIC
        presenter.voteSettings.endTime = System.currentTimeMillis() + 3000 * 86400 * 1000
        `when`(settingFragmentView.getFinalVoteSettings(voteData)).thenReturn(voteData)

        presenter.submitCreateVote()
        verify<ActivityView>(activityView).showLoadingCircle()
        verify(voteDataRepository, never()).createVote(any<VoteData>(), ArgumentMatchers.anyList()
                , any<File>(), capture(getVoteDataCallbackArgumentCaptor))
        verify<ActivityView>(activityView).hideLoadingCircle()
        verify<ActivityView>(activityView).showCreateVoteError(ArgumentMatchers.anyMap())
    }

    @Test
    fun addAndReviseAndRemoveOptionUpdateToView() {
        presenter = CreateVoteActivityPresenter(voteDataRepository
                , userDataRepository, activityView, optionFragmentView, settingFragmentView)
        val optionId = presenter.addNewOption()
        presenter.addNewOption()
        presenter.addNewOption()
        presenter.reviseOption(optionId, "newText")
        var equalText = ""
        for (option in presenter.optionList) {
            if (option.id == optionId) {
                equalText = option.title
            }
        }
        Assert.assertEquals(equalText, "newText")
        verify<OptionFragmentView>(optionFragmentView, times(3)).refreshOptions()
        val optionNumber = presenter.optionList.size
        presenter.removeOption(optionId)
        verify<OptionFragmentView>(optionFragmentView, times(4)).refreshOptions()
        Assert.assertEquals(presenter.optionList.size, optionNumber - 1)
    }

    @Test
    fun updateVoteSecurityAndEndTimeAndTitleAndUpdateToView() {
        presenter = CreateVoteActivityPresenter(voteDataRepository, userDataRepository, activityView, optionFragmentView, settingFragmentView)
        presenter.updateVoteSecurity(VoteData.SECURITY_PRIVATE)
        Assert.assertEquals(presenter.voteSettings.security, VoteData.SECURITY_PRIVATE)
        presenter.updateVoteEndTime(System.currentTimeMillis() - CreateVoteActivityPresenter.DEFAULT_END_TIME * 1000 * 86400)
        verify<ActivityView>(activityView).showHintToast(ArgumentMatchers.anyInt())
        presenter.updateVoteEndTime(System.currentTimeMillis() + (CreateVoteActivityPresenter.DEFAULT_END_TIME_MAX + 10) * 1000 * 86400)
        verify<ActivityView>(activityView).showHintToast(ArgumentMatchers.anyInt())
        presenter.updateVoteEndTime(System.currentTimeMillis() + CreateVoteActivityPresenter.DEFAULT_END_TIME_MAX * 1000 * 86400)
        verify<SettingFragmentView>(settingFragmentView).setUpVoteSettings(any<VoteData>())
        presenter.updateVoteImage(File("test"))
        Assert.assertNotNull(presenter.voteSettings.imageFile)
        presenter.updateVoteTitle("title")
        Assert.assertEquals(presenter.voteSettings.title, "title")
    }

    companion object {
        private lateinit var user: User
        private val voteCode = "CODE_123"
        private val voteData = VoteData()
        private val optionList = ArrayList<Option>()
    }
}
